crates/re_log_types/src/lib.rs

```diff
@@ -457,22 +464,38 @@ macro_rules! component_legacy_shim {
                 Ok(arrow)
             }
 
+            #[inline]
             fn try_from_arrow_opt(
                 data: &dyn arrow2::array::Array,
             ) -> re_types::DeserializationResult<Vec<Option<Self>>>
             where
                 Self: Sized,
             {
                 use arrow2_convert::deserialize::arrow_array_deserialize_iterator;
-
                 let native = arrow_array_deserialize_iterator(data)
                     .map_err(|err| {
                         re_types::DeserializationError::ArrowConvertFailure(err.to_string())
                     })?
                     .collect();
+                Ok(native)
+            }
 
+            #[inline]
+            fn try_from_arrow_opt_iter(
+                data: &dyn arrow2::array::Array,
+            ) -> re_types::DeserializationResult<Self::IterItem<'_>>
+            where
+                Self: Sized,
+            {
+                let native =
+                       <<Self as arrow2_convert::deserialize::ArrowDeserialize>::ArrayType as arrow2_convert::deserialize::ArrowArray>::iter_from_array_ref(data);
                 Ok(native)
             }
+
+            #[inline]
+            fn iter_mapper(item: Self::Item<'_>) -> Option<Self> {
+                <Self as arrow2_convert::deserialize::ArrowDeserialize>::arrow_deserialize(item)
+            }
```

Comment:
Introducing a separate mapper function from the arbitrary `Item` to a `Self` was a big part of the trick.  This lets us apply the mapper in `DataCell` instead where we're allowed to just return a `impl Iterator<Item = Option<C>`, which we can't otherwise do from the trait definition.

```
pub fn try_to_native_opt<'a, C: Component + 'a>(
    &'a self,
) -> DataCellResult<impl Iterator<Item = Option<C>> + '_> {
    Ok(C::try_from_arrow_opt_iter(self.inner.values.as_ref())?.map(C::iter_mapper))
}
```
```